src/cargo/ops/registry/publish.rs

```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
+        }
+        if Packages::Default == opts.to_publish && ws.is_virtual() {
+            bail!("the `-p` argument must be specified in the root of a virtual workspace")
+        }
     }
-    if Packages::Default == opts.to_publish && ws.is_virtual() {
-        bail!("the `-p` argument must be specified in the root of a virtual workspace")
+
+    let member_ids: Vec<_> = ws.members().map(|p| p.package_id()).collect();
+    // Check that the specs match members.
+    for spec in &specs {
+        spec.query(member_ids.clone())?;
     }
-    let member_ids = ws.members().map(|p| p.package_id());
-    // Check that the spec matches exactly one member.
-    specs[0].query(member_ids)?;
     let mut pkgs = ws.members_with_features(&specs, &opts.cli_features)?;
     // In `members_with_features_old`, it will add "current" package (determined by the cwd)
     // So we need filter
     pkgs = pkgs
         .into_iter()
         .filter(|(m, _)| specs.iter().any(|spec| spec.matches(m.package_id())))
         .collect();
-    // Double check. It is safe theoretically, unless logic has updated.
-    assert_eq!(pkgs.len(), 1);
 
-    let (pkg, cli_features) = pkgs.pop().unwrap();
-
-    let mut publish_registry = match opts.reg_or_index.as_ref() {
-        Some(RegistryOrIndex::Registry(registry)) => Some(registry.clone()),
-        _ => None,
-    };
-    if let Some(ref allowed_registries) = *pkg.publish() {
-        if publish_registry.is_none() && allowed_registries.len() == 1 {
-            // If there is only one allowed registry, push to that one directly,
-            // even though there is no registry specified in the command.
-            let default_registry = &allowed_registries[0];
-            if default_registry != CRATES_IO_REGISTRY {
-                // Don't change the registry for crates.io and don't warn the user.
-                // crates.io will be defaulted even without this.
-                opts.gctx.shell().note(&format!(
-                    "found `{}` as only allowed registry. Publishing to it automatically.",
-                    default_registry
-                ))?;
-                publish_registry = Some(default_registry.clone());
+    let just_pkgs: Vec<_> = pkgs.iter().map(|p| p.0).collect();
+    let reg_or_index = match opts.reg_or_index.clone() {
+        Some(r) => {
+            validate_registry(&just_pkgs, Some(&r))?;
+            Some(r)
+        }
+        None => {
+            let reg = super::infer_registry(&just_pkgs)?;
+            validate_registry(&just_pkgs, reg.as_ref())?;
+            if let Some(RegistryOrIndex::Registry(ref registry)) = &reg {
+                if registry != CRATES_IO_REGISTRY {
+                    // Don't warn for crates.io.
+                    opts.gctx.shell().note(&format!(
+                        "found `{}` as only allowed registry. Publishing to it automatically.",
+                        registry
+                    ))?;
+                }
             }
```

Comment:
nit: looks like switching to `infer_registry` could be broken out into a refactor commit?  Within the existing single-package publish, it should have no behavior change but its a stepping stone towards multi-package publish.

Reply:
Thanks!  This makes it really clear that that commit is actually diagnostic fix!

(technically, it could be pulled out into its own PR then... but not going to block on that)